bpo-44002: Switch to lru_cache in urllib.parse.

gpshead opened the pull request:
urllib.parse now uses functool.lru_cache for its internal URL splitting and
quoting caches instead of rolling its own like its the 90s.

The undocumented internal Quoted class is now deprecated, for removal in 3.14.
No existing users were found.

clear_cache() is kept as twisted, gevent, and our own regression test runner were
found to be using it in their test suites.

<!--
Thanks for your contribution!
Please read this comment in its entirety. It's quite important.

# Pull Request title

It should be in the following format:

```
bpo-NNNN: Summary of the changes made
```

Where: bpo-NNNN refers to the issue number in the https://bugs.python.org.

Most PRs will require an issue number. Trivial changes, like fixing a typo, do not need an issue.

# Backport Pull Request title

If this is a backport PR (PR made against branches other than `master`),
please ensure that the PR title is in the following format:

```
[X.Y] <title from the original PR> (GH-NNNN)
```

Where: [X.Y] is the branch name, e.g. [3.6].

GH-NNNN refers to the PR number from `master`.

-->


<!-- issue-number: [bpo-44002](https://bugs.python.org/issue44002) -->
https://bugs.python.org/issue44002
<!-- /issue-number -->

Automerge-Triggered-By: GH:gpshead

gpshead committed:
bpo-44002: Switch to lru_cache in urllib.parse.

urllib.parse now uses functool.lru_cache for its internal URL splitting and
quoting caches instead of rolling its own like its the 90s.

The undocumented internal Quoted class and clear_cache() APIs are now
deprecated, for removal in 3.14.

rhettinger commented on Lib/urllib/parse.py:
```diff
@@ -434,6 +433,8 @@ def _checknetloc(netloc):
                              "characters under NFKC normalization")
 
+# Cache size chosen by random die roll; prior to functools it was 20 in 1997.
+@functools.lru_cache(maxsize=99)
```
If it were me, I wouldn't specify a maxsize and just accept the default.   Also, there is *slight* advantage to making the maxsize a power of two.

rhettinger commented on Lib/urllib/parse.py:
```diff
@@ -884,6 +885,11 @@ def quote_plus(string, safe='', encoding=None, errors=None):
     string = quote(string, safe + space, encoding, errors)
     return string.replace(' ', '+')
 
+# Expectation: A typical program is unlikely to create more than 5 of these.
```
Again, I would go with the default size.  The LRU cache isn't presized; rather, it grows as needed.  So, there is no benefit in selecting a smaller size because of the expected number of cases.  The purpose of the *maxsize* to prevent excessive memory use in the extreme cases.  So we choose the size mainly based on our tolerance worst case.  Since the size of the inputs and outputs for a quoter are very small, we could tolerate thousands of entries and never notice the memory consumption.

rhettinger commented on Lib/test/test_urlparse.py:
```diff
@@ -1035,8 +1036,29 @@ def test_telurl_params(self):
+            self.assertIs(old_class, urllib.parse._Quoter)
+        self.assertEqual(len(w), 1, msg=repr(w))
+        v = sys.version_info
+        exc = PendingDeprecationWarning if v <= (3, 11) else DeprecationWarning
```
I would go directly to a DeprecationWarning.  We haven't had much luck with people ever seeing and benefitting from a PendingDeprecationWarning.

gpshead replied:
I'll defer deciding this until I do some OSS searching to see if I can even find _any_ users. (i bet not, these are obscure)

tirkarthi replied:
I can see two usages in grepping top 1000 PyPI packages.

```
$ rg -t py 'urllib.parse.*clear_cache'
twisted/src/twisted/web/test/test_http.py
17:from urllib.parse import clear_cache  # type: ignore[attr-defined]

gevent/src/gevent/testing/leakcheck.py
130:            sys.modules['urllib.parse'].clear_cache()
```

gpshead replied:
Thanks!  Based on that (and our own libregrtest use), i'll just keep clear_cache and not deprecate it.

gpshead committed:
Use the default lru_cache size.

gpshead reviewed (commented):
The test that's failing due to an unexpected BytesWarning in CI is annoying... that means we need to use `lru_cache(typed=True)` on the urlsplit decorator for safety to avoid other people who enable BytesWarning from seeing that warning.

gpshead committed:
Use a typed lru_cache on urlsplit.

serhiy-storchaka commented on Lib/urllib/parse.py:
```diff
@@ -897,14 +904,11 @@ def quote_from_bytes(bs, safe='/'):
         # Normalize 'safe' by converting to bytes and removing non-ASCII chars
         safe = safe.encode('ascii', 'ignore')
     else:
-        safe = bytes([c for c in safe if c < 128])
```
This is a de-optimization. List comprehension is faster than generator expression.

gpshead replied:
sad and understandable, but it appears true.  i restored them.

gpshead committed:
Restore the list comprehensions.

gpshead committed:
don't deprecate clear_cache().

twisted, gevent, and our own libregrtest use it in test suites.  We'll
keep it for them.

... and skip Pending on Quoter's DeprecationWarning.  No uses of that have
been found externally.

Also reoragnizes the unittest a bit to be cleaner and fixes a whitespace
problem.

akulakov commented on Lib/urllib/parse.py:
```diff
@@ -884,6 +881,11 @@ def quote_plus(string, safe='', encoding=None, errors=None):
     string = quote(string, safe + space, encoding, errors)
     return string.replace(' ', '+')
 
+# Expectation: A typical program is unlikely to create more than 5 of these.
```
I think comment is no longer needed? [Edit: on a second thought, might be good to have, please disregard.]

gpshead committed:
Merge remote-tracking branch 'upstream/main' into non-90s-urlparse-cache

gpshead committed:
Add a test for the module level __getattr__().
